core/drivers/crypto/versal/ipi.c

```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) Foundries Ltd. 2022 - All Rights Reserved
+ * Author: Jorge Ramirez <jorge@foundries.io>
+ */
+
+#include <arm.h>
+#include <initcall.h>
+#include <ipi.h>
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+#include <string.h>
+#include <tee/cache.h>
+
+#define SEC_MODULE_SHIFT 8
+#define SEC_MODULE_ID 5
+
+#define CRYPTO_API_ID(__x) ((SEC_MODULE_ID << SEC_MODULE_SHIFT) | (__x))
+
+TEE_Result versal_crypto_request(enum versal_crypto_api id,
+				 struct versal_cmd_args *arg, uint32_t *err)
+{
+	struct versal_ipi_cmd cmd = { };
+	size_t i = 0;
+
+	cmd.data[i] = CRYPTO_API_ID(id);
+	for (i = 1; i < arg->dlen + 1; i++)
+		cmd.data[i] = arg->data[i - 1];
+
+	/* src */
+	if (!arg->ibuf[0].mem.buf)
+		goto notify;
+
+	cmd.data[i++] = virt_to_phys(arg->ibuf[0].mem.buf);
+	cmd.data[i++] = virt_to_phys(arg->ibuf[0].mem.buf) >> 32;
```

Comment:
That would be undefined if `buf` is a 32-bit pointer.

Reply:
yes, need to do this everywhere - on all my other drivers (jens already mentioned. sorry)